lib/sspi/sspwin32.cc

```diff
@@ -9,9 +9,23 @@
 #include "squid.h"
 #include "base64.h"
 #include "ntlmauth/ntlmauth.h"
-#include "sspwin32.h"
+#include "sspi/sspwin32.h"
 #include "util.h"
 
+#if HAVE_WINDOWS_H && HAVE_SSPI_H
+
+// FARPROC is an exception on Windows to the -Wcast-function-type sanity check.
+// suppress the warning only when casting FARPROC
+template<typename T>
+T
+farproc_cast(FARPROC in)
```

Comment:
In C++, `foo_cast<>` means "cast to foo". This function does not cast to FARPROC. It casts _from_ FARPROC.

I would also use CamelCase for this helper name so that the readers of the code using it know that it is not a standard cast:

```suggestion
FarProcToFunction(FARPROC in)
```

or, to follow Windows naming scheme,

```suggestion
FarProcToProc(FARPROC in)
```

or, simply,

```suggestion
ToFunction(FARPROC in)
```

or

```suggestion
ToProc(FARPROC in)
```

As usual, I do not insist on any specific spelling.

Reply:
> In C++, `foo_cast<>` means "cast to foo". 

WTF? that is a strange interpretation. It does not make any sense when applied to the standard X_cast operators.

It is supposed to be  ```[mechanism]_cast<foo>(expression)``` in this case the mechanism is Windows FARPROC type alias (inheritance, interfacing, puny-type, functor, function pointer, ... whatever it is).

> This function does not cast to FARPROC. It casts _from_ FARPROC.

Indeed, it casts toward the **type** template parameter.

Reply:
> Alex: In C++, foo_cast<> means "cast to foo". This function does not cast to FARPROC. It casts from FARPROC.

> Amos: It is supposed to be `[mechanism]_cast<foo>(expression)`

Agreed. I was wrong -- I went too far in that generalization. Sorry.

I still think we should use CamelCase for this helper name so that the readers of the code using it know that it is not a standard cast, but I do not insist on this change.